File: apps/frontend/features/interests/lib/src/application/use_cases/update_interest_use_case.dart
import 'package:social_monitor_shared_kernel/social_monitor_shared_kernel.dart';

import '../../domain/entities/interest_summary.dart';
import '../commands/update_interest_command.dart';
import '../contracts/interest_catalog.dart';

final class UpdateInterestUseCase {
  const UpdateInterestUseCase(this._catalog);

  final InterestCatalog _catalog;

  Future<Result<InterestSummary>> call(UpdateInterestCommand command) {
    if (!command.scope.isValid) {
      return Future.value(
        const Result.failure(
          ValidationFailure(
            message: 'Workspace scope is required',
            code: 'interests.workspace_scope_required',
          ),
        ),
      );
    }
    if (!command.interestId.isValid) {
      return Future.value(
        const Result.failure(
          ValidationFailure(
            message: 'Interest id is required',
            code: 'interests.id_required',
            field: 'interestId',
          ),
        ),
      );
    }
    if (!command.name.isValid) {
      return Future.value(
        const Result.failure(
          ValidationFailure(
            message: 'Interest name must contain at least two characters',
            code: 'interests.name_invalid',
            field: 'name',
          ),
        ),
      );
    }
    if (!command.query.isValid) {
      return Future.value(
        const Result.failure(
          ValidationFailure(
            message: 'Interest query must contain at least two characters',
            code: 'interests.query_invalid',
            field: 'query',
          ),
        ),
      );
    }
    return _catalog.updateInterest(command);
  }
}

File: libs/monitoring/features/update-interest/update-interest.use-case.ts
import { DomainError, err, ok, type Result } from '@social-monitor/shared-kernel';

import type { InterestRepositoryPort } from '../../ports';
import { presentInterest } from '../shared/interest-presenter';
import type { UpdateInterestCommand } from './update-interest.command';
import type { UpdateInterestResult } from './update-interest.result';

type UpdateInterestFailure = DomainError | Error;

export class UpdateInterestUseCase {
  constructor(private readonly interests: InterestRepositoryPort) {}

  async execute(command: UpdateInterestCommand): Promise<Result<UpdateInterestResult, UpdateInterestFailure>> {
    const validationFailure = validateCommand(command);
    if (validationFailure !== null) {
      return err(validationFailure);
    }

    const interest = await this.interests.findById({
      tenantId: command.tenantId,
      workspaceId: command.workspaceId,
      interestId: command.interestId,
    });
    if (interest === null) {
      return err(new DomainError('resource.not_found', 'Interest not found', { interestId: command.interestId }));
    }

    const duplicate = await this.interests.findByName({
      tenantId: command.tenantId,
      workspaceId: command.workspaceId,
      name: command.name,
    });
    if (duplicate !== null && duplicate.toSnapshot().id !== command.interestId) {
      return err(new DomainError('operation.conflict', 'Interest name is already used in this workspace', {
        name: command.name.trim(),
      }));
    }

    const updated = interest.updateDetails({
      name: command.name,
      query: command.query,
    });
    await this.interests.save(updated);

    return ok(presentInterest(updated));
  }
}

const validateCommand = (command: UpdateInterestCommand): DomainError | null => {
  if (command.interestId.trim().length === 0) {
    return new DomainError('validation.failed', 'Interest id is required');
  }
  if (command.name.trim().length < 2) {
    return new DomainError('validation.failed', 'Interest name must contain at least 2 characters');
  }
  if (command.query.trim().length < 2) {
    return new DomainError('validation.failed', 'Interest query must contain at least 2 characters');
  }
  return null;
};
